File: 2015/haskell/Day03.hs
module Day03 where

import Text.Printf(printf)
import Data.List

main = do
  input <- readFile "../inputs/03.txt"
  printf "part 1: %d\n" $ part1 input
  -- printf "part 2: %d\n" $ part2 input

part1 input = length $ nub $ positions input
part2 input = 0

-- For both parts --
direction :: Char -> [Int, Int]
direction c
  | c=='>' = [ 1,  0]
  | c=='<' = [-1,  0]
  | c=='^' = [ 0,  1]
  | c=='v' = [ 0, -1]

nextPosition :: [Int, Int] -> Char -> [Int, Int]
nextPosition position arrow = zipWith (+) position (direction arrow)

xFilter :: Char -> Int
xFilter c 
  | c=='>' = 1
  | c=='<' = -1
  | otherwise = 0

yFilter :: Char -> Int
yFilter c
  | c=='^' = 1
  | c=='v' = -1
  | otherwise = 0

xMovements :: String -> [Int]
xMovements s = map xFilter s

yMovements :: String -> [Int]
yMovements s = map yFilter s

xPositions :: String -> [Int]
xPositions s = [0] ++ (scanl1 (+) $ xMovements s)

yPositions :: String -> [Int]
yPositions s = [0] ++ (scanl1 (+) $ yMovements s)

positions :: String -> [(Int,Int)]
positions s = zip (xPositions s) (yPositions s)

-- For part 1 --

-- For part 2 --

File: 2015/python/day03.py

def read_input():
  with open('../inputs/03.txt') as f:
    return f.read()

def part1(s):
  position = (0,0)
  a = [position]
  for c in s:
    x = position[0]
    y = position[1]
    if c == '^':
      y += 1
    elif c == '>':
      x += 1
    elif c == 'v':
      y -= 1
    elif c == '<':
      x -= 1
    position = (x,y)
    a.append(position)
  print(a)
  return len(set(a))

def part2(s):
  santa = (0,0)
  robo_santa = (0,0)
  a = [santa, robo_santa]
  for i,c in enumerate(s):
    if i % 2 == 0:
      x = santa[0]
      y = santa[1]
    else:
      x = robo_santa[0]
      y = robo_santa[1]
    if c == '^':
      y += 1
    elif c == '>':
      x += 1
    elif c == 'v':
      y -= 1
    elif c == '<':
      x -= 1
    if i % 2 == 0:
      santa = (x,y)
      a.append(santa)
    else:
      robo_santa = (x,y)
      a.append(robo_santa)
  return len(set(a))

def main():
  s = read_input()
  print(f'part 1: {part1(s)}')
  print(f'part 2: {part2(s)}')

if __name__ == "__main__":
  main()
